sequencing/sequencer.go

```diff
@@ -277,57 +576,96 @@ func getRemainingSleep(start time.Time, blockTime time.Duration, sleep time.Dura
 	return remaining + sleep
 }
 
-func hashSHA256(data []byte) []byte {
-	hash := sha256.Sum256(data)
-	return hash[:]
-}
-
 // SubmitRollupTransaction implements sequencing.Sequencer.
-func (c *Sequencer) SubmitRollupTransaction(ctx context.Context, rollupId []byte, tx []byte) error {
-	if c.rollupId == nil {
-		c.rollupId = rollupId
-	} else {
-		if !bytes.Equal(c.rollupId, rollupId) {
-			return ErrorRollupIdMismatch
-		}
+func (c *Sequencer) SubmitRollupTransaction(ctx context.Context, req sequencing.SubmitRollupTransactionRequest) (*sequencing.SubmitRollupTransactionResponse, error) {
+	if !c.isValid(req.RollupId) {
+		return nil, ErrInvalidRollupId
 	}
-	c.tq.AddTransaction(tx)
-	return nil
+	err := c.tq.AddTransaction(req.Tx, c.db)
+	if err != nil {
+		return nil, fmt.Errorf("failed to add transaction: %w", err)
+	}
+	return &sequencing.SubmitRollupTransactionResponse{}, nil
 }
 
 // GetNextBatch implements sequencing.Sequencer.
-func (c *Sequencer) GetNextBatch(ctx context.Context, lastBatchHash sequencing.Hash) (*sequencing.Batch, time.Time, error) {
+func (c *Sequencer) GetNextBatch(ctx context.Context, req sequencing.GetNextBatchRequest) (*sequencing.GetNextBatchResponse, error) {
+	if !c.isValid(req.RollupId) {
+		return nil, ErrInvalidRollupId
+	}
 	now := time.Now()
-	if c.lastBatchHash == nil {
-		if lastBatchHash != nil {
-			return nil, now, errors.New("lastBatch is supposed to be nil")
-		}
-	} else if lastBatchHash == nil {
-		return nil, now, errors.New("lastBatch is not supposed to be nil")
-	} else {
-		if !bytes.Equal(c.lastBatchHash, lastBatchHash) {
-			return nil, now, errors.New("supplied lastBatch does not match with sequencer last batch")
-		}
+	c.lastBatchHashMutex.RLock()
+	lastBatchHash := c.lastBatchHash
+	c.lastBatchHashMutex.RUnlock()
```

Comment:
What's the rationale for not just holding the mutex for the entire operation? 
This method doesn't seem to be doing anything too long live, so just keeping it all atomic under one mutex lock feels reasonable and would avoid potential edge cases of the hashes being updated mid method call.

Reply:
sure. no harm in keeping the lock.